crypto/key.go

```diff
@@ -282,7 +282,20 @@ func UnmarshalPublicKey(data []byte) (PubKey, error) {
 		return nil, ErrBadKeyType
 	}
 
-	return um(pmes.GetData())
+	pk, err := um(pmes.GetData())
+	if err != nil {
+		return nil, err
+	}
+
+	switch tpk := pk.(type) {
```

Comment:
:scream:

Reply:
Do we need to cache the entire protobuf or can we just cache (and compare) the raw data.

Reply:
if we do that, then we end up still having to create a whole protobuf every time we call bytes. The problem here is with our interfaces, we have asymetry between whats given to the key to unmarshal itself, and what we expect it to give us when we ask it for its bytes.

Reply:
> if we do that, then we end up still having to create a whole protobuf every time we call bytes.

Is that a common operation (except when comparing)?